sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/BeginInvokePairedWithEndInvoke.cs

```diff
@@ -0,0 +1,230 @@
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static readonly ISet<SyntaxKind> ParentDeclarationKinds = new HashSet<SyntaxKind>
+        {
+            SyntaxKind.AnonymousMethodExpression,
+            SyntaxKind.ClassDeclaration,
+            SyntaxKind.CompilationUnit,
+            SyntaxKind.ConstructorDeclaration,
+            SyntaxKind.DelegateDeclaration,
+            SyntaxKind.DestructorDeclaration,
+            SyntaxKind.InterfaceDeclaration,
+            SyntaxKind.MethodDeclaration,
+            SyntaxKind.ParenthesizedLambdaExpression,
+            SyntaxKind.SimpleLambdaExpression,
+            SyntaxKind.StructDeclaration
+        }.ToImmutableHashSet();
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+                {
+                    var invocation = (InvocationExpressionSyntax)c.Node;
+                    var semantic = c.SemanticModel;
+                    var methodSymbol = GetMethodSymbol(invocation, semantic);
+                    if (methodSymbol?.Name == "BeginInvoke" && IsDelegate(methodSymbol))
+                    {
+                        var callbackArg = GetCallbackArg(invocation);
+                        if (callbackArg != null &&
+                            (IsNullLiteral(callbackArg) || !IsCallbackMayContainEndInvoke(callbackArg, semantic)) &&
+                            !IsParentMethodContainsEndInvoke(invocation, semantic))
+                        {
+                            var location = ((SyntaxToken)invocation.GetMethodCallIdentifier()).GetLocation();
+                            c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, location));
+                        }
+                    }
+                },
+                SyntaxKind.InvocationExpression);
+        }
+
+        private static bool IsParentMethodContainsEndInvoke(SyntaxNode node, SemanticModel semantic)
+        {
+            var parentContext = node.AncestorsAndSelf()
+                .FirstOrDefault(ancestor => ParentDeclarationKinds.Contains(ancestor.Kind()));
+            return GetEndInvokeList(parentContext, semantic).Count > 0;
+        }
+
+        private ExpressionSyntax GetCallbackArg(InvocationExpressionSyntax invocationExpression)
+        {
+            var callbackArgPos = invocationExpression.ArgumentList.Arguments.Count - 2;
+            var callbackArg = GetArgumentExpressionByNameOrPosition(invocationExpression, "callback", callbackArgPos)
+                ?.RemoveParentheses();
+            return callbackArg;
+        }
+
+        private static bool IsCallbackMayContainEndInvoke(SyntaxNode callbackArg, SemanticModel semantic)
```

Comment:
regarding the name, I think you can skip the `Is` because it's a bit unnatural - `CallbackContainsEndInvoke`  should suffice

Reply:
Ok but I used `CallbackMayContainEndInvoke` and not not `CallbackContainsEndInvoke` because:
* false means we resolved callback code and didn't find `EndInvoke`
* true means we found `EndInvoke` or didn't resolve the callback code.

Reply:
ok, then it's worth adding these 2 points in the method comment